components/eamxx/src/control/atmosphere_driver.cpp

```diff
@@ -1283,6 +1288,56 @@ void AtmosphereDriver::set_initial_conditions ()
     }
   }
 
+  // Compute IC perturbations of GLL fields (if requested)
+  using vos = std::vector<std::string>;
+  const auto perturbed_fields = ic_pl.get<vos>("perturbed_fields", {});
+  const auto num_perturb_fields = perturbed_fields.size();
+  if (num_perturb_fields > 0) {
+    m_atm_logger->info("    [EAMxx] Adding random perturbation to ICs ...");
+
+    EKAT_REQUIRE_MSG(m_field_mgrs.count("Physics GLL") > 0,
+                     "Error! Random perturbation can only be applied to fields on "
+                     "the GLL grid, but no Physics GLL FieldManager was defined.\n");
+    const auto& fm = m_field_mgrs.at("Physics GLL");
+
+    // Setup RNG. If no random seed is requested in
+    // initial_conditions params, generate one using rand().
+    const auto seed = ic_pl.get<int>("perturbation_random_seed", rand());
+    std::mt19937_64 engine(seed);
+    m_atm_logger->info("      For IC perturbation, random seed: "+std::to_string(seed));
+
+    // Get perterbation limit. Defines a range [-perturbation_limit, perturbation_limit]
+    // for which the perturbation value will be randomly generated from.
+    const auto perturbation_limit = ic_pl.get<Real>("perturbation_limits", 0.001);
+
+    // Define a level mask using reference pressure and the pressure_lower_bound parameter.
+    // This mask dictates which levels we apply a perturbation.
+    const auto gll_grid = m_grids_manager->get_grid("Physics GLL");
+    gll_grid->get_geometry_data("hyam").sync_to_host(), gll_grid->get_geometry_data("hybm").sync_to_host();
+    const auto hyam = gll_grid->get_geometry_data("hyam").get_view<const Real*, Host>();
+    const auto hybm = gll_grid->get_geometry_data("hybm").get_view<const Real*, Host>();
+    constexpr auto ps0 = physics::Constants<Real>::P0;
+    const auto pressure_lower_bound = ic_pl.get<Real>("perturbation_pressure_lower_bound", 1050.0);
+    auto pressure_mask = [&] (const int ilev) {
+      const auto pref = (hyam(ilev)*ps0 + hybm(ilev)*ps0)/100; // Reference pressure ps0 is in Pa, convert to milibar
```

Comment:
typo in comment: milibar->millibar

Reply:
More importantly though - why are we reconstructing hydrostatic pressure from hyam, hybm? Do we not have the full pressure at this point? I'm a bit worried about this because we hope to eventually start using non-hydrostatic pressure for physics, so it would be better to just use pressure where we can. But if doing that isn't convenient right now, we can just leave this as something to fix if/when we switch to full pressure. Or maybe we never change this because it's just in initialization and is approximately correct anyways.

Reply:
This is just how EAM constructed it. At this point we would have the pressure from the IC file if that is better to use.

Reply:
Using the actual pressure from the IC file seems better BUT would give different number of masked levs for each different col. Which could cause weird stuff to happen if, for example, we only want to perturb below 700mb but we have a Mt Everest grid cell whose surface is higher (in z) than 700mb so no cells get perturbed at all. I think that's fine for our uses, but I'm curious what @bogensch thinks.

Reply:
For DP, the pressure will be the same in each column, but obviously not true in the general case.

Reply:
> For DP, the pressure will be the same in each column, but obviously not true in the general case.
I'm leaning towards using the current impl a) out of laziness and b). because I think what we really want is perturbations in the first N levels rather than below a strict pressure cutoff. Specifying a pressure is just shorthand for the number of levels. Though maybe we should just directly specify the number of levels then? @bogensch ?